### ISSUE TITLE
Missed optimization comparing TypeId arrays

### ISSUE THREAD

**Issue by AngelicosPhosphoros** [rust-lang/rust#84253]
**Edit:** Look nikic example [below](https://github.com/rust-lang/rust/issues/84253#issuecomment-821560496).

I tried this code:

```rust
use std::any::TypeId;

trait CheckEqTuple{
    type OutSliceType: AsRef<[TypeId]> + AsMut<[TypeId]>;
    fn get_unsorted_ids()->Self::OutSliceType;
    fn get_quick_sorted_ids()->Self::OutSliceType{
        let mut ret = Self::get_unsorted_ids();
        ret.as_mut().sort_unstable();
        ret
    }
    fn get_slow_sorted_ids()->Self::OutSliceType{
        let mut ret = Self::get_unsorted_ids();
        ret.as_mut().sort();
        ret
    }
}

impl<A: 'static, B: 'static> CheckEqTuple for (A, B){
    type OutSliceType = [TypeId; 2];
    fn get_unsorted_ids()->Self::OutSliceType{
        [TypeId::of::<A>(), TypeId::of::<B>()]
    }
}

fn cmp_unord<A: CheckEqTuple, B: CheckEqTuple>()->bool{
    A::get_unsorted_ids().as_ref()
     == B::get_unsorted_ids().as_ref()
}

fn cmp_ord_q<A: CheckEqTuple, B: CheckEqTuple>()->bool{
    A::get_quick_sorted_ids().as_ref() 
        == B::get_quick_sorted_ids().as_ref()
}

fn cmp_ord_s<A: CheckEqTuple, B: CheckEqTuple>()->bool{
    A::get_slow_sorted_ids().as_ref()
     == B::get_slow_sorted_ids().as_ref()
}

pub fn cmp_same()->bool{
    cmp_unord::<(i32, bool), (i32, bool)>()
}

pub fn cmp_diff_order()->bool{
    cmp_unord::<(bool, i32), (i32, bool)>()
}

pub fn cmp_ord()->bool{
    cmp_ord_q::<(bool, i32), (i32, bool)>()
}

pub fn cmp_ord2()->bool{
    cmp_ord_s::<(bool, i32), (i32, bool)>()
}

```

I expected to see this happen: 

This code should compile to `return true` or `return false`.
```asm
example::cmp_same:
        mov     al, 1
        ret

example::cmp_diff_order:
        xor     eax, eax
        ret

example::cmp_ord:
        mov     al, 1
        ret
```

Instead, this happened:

asm contains useless operations
```asm
example::cmp_same:
        sub     rsp, 32
        movabs  rax, -8661621401413125213
        mov     qword ptr [rsp + 8], rax
        mov     al, 1
        add     rsp, 32
        ret

example::cmp_diff_order:
        sub     rsp, 32
        movabs  rax, -5015437470765251660
        mov     qword ptr [rsp + 8], rax
        xor     eax, eax
        add     rsp, 32
        ret

example::cmp_ord:
        sub     rsp, 32
        movabs  rax, -5015437470765251660
        mov     qword ptr [rsp + 8], rax
        mov     al, 1
        add     rsp, 32
        ret
```

### Meta

`rustc --version --verbose`:
```
rustc 1.51.0 (2fd73fabe 2021-03-23)
```

### Proposed fix
If I add additional SROA flag, useless code is eliminated. I think, we need to add another SROA to pipeline somewhere before last instcombine and dead code elimination.

Here [godbolt link](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAM1QDsCBlZAQwBtMQBGAFlJvoCqAZ0wAFAB4gA5AAYppAFZdSrZrVDEArkIKkR7ZATx1KmWugDCqVpoC2tEN15n0AGTy1MAOTsAjTMQgAKwAzKQADqhChMa0Vjb2jryR0UZ07p4%2Btv6BoXqYBmm0DATMxATxdg5O%2BYWxJWUEGd5%2BAcFhOo2ViTWd5c1ZOe0AlHqomsTIHFLamADUOuggIGoAnssAKqvhmACS6ACkIQBCBzIAgmfnBMTMhHMWCJjIANbYAI4bmuHsBwDspwuc2BcwI23mAHlNIxWHgplsdiA5uchAAlTBUI4WA5BY4IvaHIIAESO2DmBwATMdkUIALLQrE4vHg/Y4kkhbBHQHnEFzKi0ObATAEAD6mloQhIBEw6BFeHQQggwwAtKSGAUqMsoTC4Zh8Vyrrz%2BYLhSL3po4S8RZLyjK5QqlaqOerWJqQNqGLD4eD/tzebz2AQ5rZoXNiMLySEiXMXW6haLxTbpbL5Yrhgagf7geGCAA6ZhCEUhghK3NJsUS0q%2BdhKjM8rNh4WGkH/EmZ4HG%2BPW1ioADu1qldtTjrVGq10M9uv1AObWcDwdDOcj0djyy7icHKYd6ZOs/9OfzheLpaTtd37f3TYvrauN4uVzwth%2BWPOSLAUk6RmQpDmxzfH9KL9SQeJ5Xg%2BL4fnmGhiDmCBzh/Y50xnC8wR2OYPS9PVwWXclcXxVkTjmCk2TrI0BXXCVN3tNMnWwVd3QnTDpz9BsmXwpYQFQN0X1JJUf3Y5YuOWLFTg5WtiT3O9Lj%2BNtLguY1kCfCsSEOEILFfEDnjeT5vnYBCkUeLTwN0zBeJVUlfFQaxfT3V8QAopMh23Q8RXDKhawvZcjmjP97NNDdbS3NMXLcjzpNkq4FKUlSzRfAzQO0iC9N/eKjJ0yCzNoyzrOQ%2BtgTsrtzUtAdAuo0sC1cjElXJTyWyjbyUr80UiteErkzK4YQqq9N7xk295IFRTwhFGKhDizSwPS5K/wmxKTMyiyrNYGyLwK00hB7ftHKC8rC1Cnq8uBbyGt8rsNr7NqnOCir9tvPr7wucJNF8PlBqUoRmFsTARw5bLlty3khuU4gOKxCA8BCCkfz%2B0ZYIhqG5hhsy7oix7nteuYgfwKgqBGkGAh%2B7A/pWw6gfFFThLUiA/p/eHYfByHoaWndsDCqSriel6ouGlTCeJgGQSBmL3kpixqaW2nIfp%2BGmesFm2fuuTzk5jGhZBik%2BaWknAeikHrVF8XrElilpcZxHmeR3rUeV9HjVsZhVn8EVwjUOFawF4E8CoTHdfQDWkJYrMcwmWhSJbRXeRd2g4TASBKQpAAJAoe3jg7WykUZWGkIJ5AcWR5FQaRUW0INJQmKZyQpEJOHkAhpDkYZRheEAQhCXNW47zuu4ANhUaRuHkWwuBkGRSDzuRSELqR5CEEBR7r/PRjgWAYEQFBUCfPB2DICgIDQTft5AYhOCCThR6oLfpWIWfqfr%2BRfA8Mp1ikGvSH3r76AhWhWGfiesHt9Q7A76kHwOGQweAABumBZ751IJgcQzxoTTFfh4aUWdYGwl8LcYgqwrBYGATcR80ga6jH4IwFgQCeB8DoAQYQYhJCwKUJwFQagNAlxUHgXws9ICjFQOEIoMDlQWDmHwggyp2BQOWlGEIcwhFzBdkIEQY16oUgpEIYgqBmBzA8DofeD9PCCh7L4Ng6ApjxxngUZ4RRTDmG6A4ZhLgBitECMwlIMQ6B2OUG4ooTjshtGYfoKx9RSjlE8QEyx4C6ANH6B4FofiXF6BCRUawVRlB9CaLEwY/jRhl0mBwTgmds652AVPLQui5icFzKfXMMhYK4EICQSu1cfxWAPgEJpBS5jFx0LXO%2BjdSBPGYFgQIfFm5BFHuggepAh7jLHiU6QM856kAXg3Ugy81773CFvAI5BKCbO2YEY%2Bp9z6XwCDfXwwD9FP2IfId%2BZgCBfx/sA/%2BrCgGwNAUEqBMCJ7wMQdKG55B6AFGAZg7BuCMDTAnoQoeL8M7UPoEwNg%2BTeBkLoRIYBSgoaqHUEfdhmDuF8VEbEQRwjRHiMwJI7yIQLF1BMBAFwYS4HmF8UMVxUR3FxBSYkNlqRYgsv8bUIJUSkmMsCZE4oST%2BUJPSaKyVmTnFcByeMPJiq%2B5SBznM2BpSS7IAqVUyptSID1KIDBSkzSHgby2dvJpJsukl16YvUYgzhmUEKVIKZMzR7jwLgsvQSyVkZybi3CkuZu4AE4AAcPAw3dwjRSbgMgo1/DVdSzVE8p4OtWesiASBsZUF2dQGhiLKEopoWihhE9WAIFnswqtQg82oQ4KPYg1aQBQxbfWr2VBG1zzdRq71k9pBEi7WGHVerqm1IgUIUd5TKkTszYG0gzdW65hjRG0%2B4aI3cApEELdYa1VTIHRmv188%2BluopMUrVvqA39KgdfWIjggA%3D%3D).

**Comment by nikic** [comment]
A simpler starting point that already doesn't optimize: https://godbolt.org/z/rvMah5b5o

**Comment by AngelicosPhosphoros** [comment]
@nikic Do you know if rustc provides some custom LLVM passes to pipeline? After skimming of source, it looks like it just uses default optimization levels from LLVM.

**Comment by nikic** [comment]
@AngelicosPhosphoros Yes, rust uses only the default optimization pipeline.

In this case, I believe the problem lies with unrolling. After some cleanup, we have this IR going into loop-unroll-full: https://gist.github.com/nikic/9516dede0acedd8d320e9d067559e667

If I understand correctly, it is not unrolled fully because we know small upper bound on the tripcount (3), but the target (X86) does not enable UpperBound unrolling. It does work with `-mtriple=arm`, which enables it. This was kinda surprising to me, given LLVMs usual propensity to vectorize and unroll everything beyond any reasonable bounds. There's this comment in the implementation:

```
  // We can unroll by the upper bound amount if it's generally allowed or if
  // we know that the loop is executed either the upper bound or zero times.
  // (MaxOrZero unrolling keeps only the first loop test, so the number of
  // loop tests remains the same compared to the non-unrolled version, whereas
  // the generic upper bound unrolling keeps all but the last loop test so the
  // number of loop tests goes up which may end up being worse on targets with
  // constrained branch predictor resources so is controlled by an option.)
  // In addition we only unroll small upper bounds.
```

Which makes me think that this option should probably enabled on X86, which is certainly not a system with "constrained branch predictor resources".

Though there is another important factor here: What the constant upper bound doesn't capture, is the fact that in this case, there's two loop exits. One of the exits will be replicated by unrolling. However, the other one is fully determined by unrolling. These exits will be eliminated, together with the induction variable.

This means that the loop unrolling heuristic is inconsistent: If the loop had an interior (non-exiting) branch, then it *would* get fully unrolled, even though that replicates the exact same number of branches. It's just a difference between whether those branches are exiting or not.

**Comment by nikic** [comment]
Looking a bit further, there is this code: https://github.com/llvm/llvm-project/blob/ebc6608fb79057eaed27435d62d5dea0979bd9d3/llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp#L1099-L1108 For exact trip counts, LoopUnroll will actually not use the loop trip count, but rather the trip count of a single exit *or* the latch exit. This is really the case we want to fall into. Unfortunately, in our case the latch exit is the comparison of the loads (which has an unknown exit count) rather than the comparison of the induction variant (which is known).

There's two ways that could be addressed: The LoopUnroll logic could be more general about which exit count it uses. Or LoopRotate could be adjusted to perform a rotation in this case despite already having an exiting latch (there is already a heuristic for that, but it does not account for this case).

**Comment by nikic** [comment]
This issue should be addressed by https://reviews.llvm.org/D102982 (plus about a dozen preparatory patches to make loop unrolling robust against multiple exits).

**Comment by nikic** [comment]
This is fixed by #87570. New IR folds to true/false:

```
@_ZN5test28cmp_same17hf172c8bcc9b7859cE = unnamed_addr alias i1 (), i1 ()* @_ZN5test27cmp_ord17ha9ee7b72078df8e5E
@_ZN5test28cmp_ord217h23f0a5da1c84ecaaE = unnamed_addr alias i1 (), i1 ()* @_ZN5test27cmp_ord17ha9ee7b72078df8e5E

; test2::cmp_diff_order
; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn
define zeroext i1 @_ZN5test214cmp_diff_order17hd0c16acee48eb39bE() unnamed_addr #0 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality {
start:
  ret i1 false
}

; test2::cmp_ord
; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone uwtable willreturn
define zeroext i1 @_ZN5test27cmp_ord17ha9ee7b72078df8e5E() unnamed_addr #0 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality {
start:
  ret i1 true
}
```
